### PR TITLE
Add option to reindex backend search data that has not been reindexed yet

### LINKED ISSUE #8077: Parallelize SearchBackendReindexCommand
Currently for big Pimcore installations the SearchBackendReindexCommand is un-usable due to doctrine memory leak problems as all Pimcore elements are loaded within one command. I tried it for an installation with 1 million assets and 4 million data objects.

As improvement it should perfectically be possible to parallelize this command using the Webmozarts console parallelization trait.

### PR DESCRIPTION
**Pull request by alyna**
The SearchBackendReindexCommand can't be used because of memory issues. Each time I run _bin/console pimcore:search-backend-reindex_ after a while the process stops and I get the error:
 _PHP Fatal error: Allowed memory size of 1048576000 bytes exhausted (tried to allocate 1769472 bytes) in /var/www/html/vendor/symfony/error-handler/Resources/views/logs.html.php on line 38_

I tested it on a data with 455000 assets, 3700 documents, 3900000 objects.

In this solution I added an option so that when the command is run the whole data is reindexed:  _bin/console pimcore:search-backend-reindex --full=1_
If the command is run without the _full_ parameter, then it will reindex only the data that has not been reindexed yet.

## Changes in this pull request  
Resolves #8077

## Additional info  
I also found this merge request https://github.com/pimcore/pimcore/pull/10423 but I still get the out of memory error, even with this fix.

### CODE REVIEW

**Comment by mcop1** [comment]
@alyna Thanks for the pr, but this is changing the default behavior, right? Default right now is to reindex everything, with your pr the default would be to reindex the missing ids only. We would like to keep the default behavior as it is. 

Could you please change the logic in your pr, so that you have to pass a parameter to index the missing ids only?

**Comment by CodeBrauer** [comment]
By indexing only missing IDs with this solution, how would the table be optimized afterwards? (Running this in a cronjob)

File: bundles/CoreBundle/src/Command/SearchBackendReindexCommand.php [hunk 1]
```diff
@@ -64,6 +75,10 @@ protected function execute(InputInterface $input, OutputInterface $output): int
                     DataObject::OBJECT_TYPE_VARIANT,
                 ]);
             }
+            // Load the data that has already been processed
+            if (!$isFull && $ids = $db->fetchFirstColumn("SELECT id FROM search_backend_data WHERE maintype = '". $type ."' ")) {
+                $list->setCondition($type === 'object' ? 'o_id' : 'id' .' NOT IN (:oids)', ['oids' => $ids]);
```
Comment (Countryen) [inline bundles/CoreBundle/src/Command/SearchBackendReindexCommand.php:80]: Maybe this is wrong?
Should this resolve to `'o_id'` or `'o_id NOT IN (:oids)'` for type 'object'?

I think it should be changed to:
`$list->setCondition(($type === 'object' ? 'o_id' : 'id') .' NOT IN (:oids)', ['oids' => $ids]);`
(see the extra braces)

We did this and without the proposed change he restarted all objects every time.

Comment (jdreesen) [reply bundles/CoreBundle/src/Command/SearchBackendReindexCommand.php:80]: As this is for Pimcore 11... wasn't the `o_`-prefix recently removed?

/edit: oh, no, it's not ready yet: #13308

**Comment by Countryen** [comment]
I've had a very similar issue when upgrading to 10.5. Thanks for your PR, I have created a separate command with this content until this is merged into pimcore.

Additionally to the memory issue, the command creates temporary files (var/tmp) and this was too much file size for our setup (~3 Million entries).

Maybe it would also help if the command could be run with something like --async, creating SearchBackendMessages instead of doing it synchronous.

**Comment by mcop1** [comment]
> I've had a very similar issue when upgrading to 10.5. Thanks for your PR, I have created a separate command with this content until this is merged into pimcore.
> 
> Additionally to the memory issue, the command creates temporary files (var/tmp) and this was too much file size for our setup (~3 Million entries).
> 
> Maybe it would also help if the command could be run with something like --async, creating SearchBackendMessages instead of doing it synchronous.

@Countryen Please open a new issue/improvement for your suggestion, otherwise it can get overlooked easily

File: bundles/CoreBundle/src/Command/SearchBackendReindexCommand.php [hunk 2]
```diff
@@ -64,6 +75,10 @@ protected function execute(InputInterface $input, OutputInterface $output): int
                     DataObject::OBJECT_TYPE_VARIANT,
                 ]);
             }
+            // Load the data that has already been processed
+            if (!$isFull && $ids = $db->fetchFirstColumn("SELECT id FROM search_backend_data WHERE maintype = '". $type ."' ")) {
```
Comment (brusch) [inline bundles/CoreBundle/src/Command/SearchBackendReindexCommand.php:79]: Using this query without a limit could potentially lead to memory issues, I'd suggest to do them in chunks/batches?

**Review by brusch — CHANGES_REQUESTED** [review]

File: bundles/CoreBundle/src/Command/SearchBackendReindexCommand.php [hunk 3]
```diff
@@ -34,6 +35,13 @@ protected function configure()
         $this
             ->setName('pimcore:search-backend-reindex')
             ->setAliases(['search-backend-reindex'])
+            ->addOption(
+                'full',
+                'f',
+                InputOption::VALUE_OPTIONAL,
```
Comment (BlackbitDevs) [inline bundles/CoreBundle/src/Command/SearchBackendReindexCommand.php:41]: Wouldn't `VALUE_NONE` be better? The CLI call then would look like `bin/console pimcore:search-backend-reindex --full`.

**Comment by fashxp** [comment]
Closing due to inactivity. Feel free to ping us for reopening when comments and merge conflicts are resolved.

**Pull request closed without merging**